tools/autograd/derivatives.yaml

```diff
@@ -1206,8 +1206,8 @@
   self: log_sigmoid_double_backward(grad * grad_output, self)
 
 - name: _log_softmax_backward_data(Tensor grad_output, Tensor output, int64_t dim, Tensor self)
-  grad_output: grad - (grad * output.exp()).sum(dim, true)
-  self: log_softmax_double_backward(grad, grad_output, dim, output).type_as(self)
+  grad_output: grad.type_as(output) - (grad.type_as(output) * output.exp()).sum(dim, true)
```

Comment:
Could you use `.to(output.dtype())` instead? I know it's not a problem here, but `type_as` ignores devices index and I think we should generally avoid using it.

Reply:
Also replace type_as in clamp derivative.

Reply:
Thanks!

Reply:
I also wonder if we can avoid calculating `grad.type_as(output)` twice, but I guess that requires putting this into a function, and can be left as a future optimization.

Reply:
Yeah, I've noticed it, but given that it's really a corner case (in most cases type_as will be a no-op), decided to leave it as is for now

Reply:
SGTM